### ISSUE TITLE
Unbounded memory allocation in bubble_neighbors

### ISSUE THREAD

**Issue by MLucient** [mchalela/GriSPy#8]
First off - thanks so much for GriSPy: it's working very well for my use case (mass manipulations of 3D reconstructions)! 
Anyhow, I wanted to let you know that there is some unbounded memory allocation in the bubble_neighbors function that can cause an unexpected hard crash, which I suspect can probably be fixed without too much trouble.

Crashing use case:
I have a 3D reconstruction file with ~20M 3D points, and I want to find all points within {x} distance of a smaller set (~1K) points. (So that I can cut out all other points). Grid-search is a great improvement for me performance-wise on other solutions I'd tried. But if I run the following, where GSP is the initialized grid with ~20M points, and nvm_df is my positions pandas dataframe with the 1K centers to check:
distances, indices = gsp.bubble_neighbors(np.vstack(nvm_df['positions']), distance_upper_bound=8)

I can watch my memory usage climb until a crash (64GB workstation, so this caught me off guard!)

There is a very simple work-around: I can run this same grid search in a loop (between 50-100 centers at a time) and manually keep a running tally of the found indices myself in the loop. In that case memory use cyclically grows and then is returned with each finished gsp.bubble_neighbors invocation. There isn't a memory leak proper, since the loop version works, but the naive version crashing is a surprise since the bubble_neighbors function already accepts an array of centers. 

Please let me know if I can further help diagnose! Alternately, if you'd like a PR, I could try to take a look.

**Comment by mchalela** [comment]
Hi @MLucient! I'm glad you are finding grispy useful :)
I've tried a naive example with the same (aproximate) numbers you are using, 20 million 3d points and 1 thousand centers. In a 16 gb pc I had no problems. So, I'm thinking that maybe the problem is in the vstack call. Can you try to split in two the bubble_neighbors call to this:

```python
pos = np.vstack(nvm_df['positions'])
distances, indices = gsp.bubble_neighbors(pos, distance_upper_bound=8)
```
And check which call produces the crash. If it still happens in the bubble, can you post here the output of `nvm_df['positions'].head()`


This was my example, this is an edited version of the example script:

```python
import numpy as np
from grispy import GriSPy

# Create random points and centres
Npoints = 20 * 10**6
Ncentres = 1000
dim = 3
Lbox = 100.0

np.random.seed(42)
data = np.random.uniform(0, Lbox, size=(Npoints, dim))
centres = np.random.uniform(0, Lbox, size=(Ncentres, dim))

# Grispy params
upper_radii = 8.0

# Build the grid with the data
gsp = GriSPy(data, periodic=periodic)

# Query for neighbors within upper_radii
bubble_dist, bubble_ind = gsp.bubble_neighbors(
    centres, distance_upper_bound=upper_radii
)
```

**Comment by MLucient** [comment]
Hi Martin - thanks for your response! I'm pretty sure it's not the vstack call (since that's just a numpy reshaping), but I factored it out to be earlier in the code as a precaution. And the resulting position shape of the numpy array is (700, 3) for the example I'm sending.

I'm figuring the easiest thing to do is give you a better example of what I'm seeing, so I modified an instance of my use case to help iterate on for a stand-alone script, and am sharing data files here. The former is the 20M points (+ some other data that doesn't end up used), and is a big file (about .5 GB), the second is the camera position data that has the centers I'm interested in searching arond, and should be a small file (< 100 KB).

These should be open-shared, please let me know if they aren't
test_data.csv -> https://drive.google.com/file/d/10UHod-p9V5uDOslWVgjagDKLw85PLVNL/view?usp=sharing
nvm_data.csv -> https://drive.google.com/file/d/1fAsChVuD18MkONKARUlRNMib92P3H82c/view?usp=sharing

The standalone script is here - should be pretty easy to modify by toggling the "step_size" variable to change how many centers GriSPy is searching for in each iteration. 

```
from itertools import count, chain
import numpy as np
import pandas as pd
from grispy import GriSPy as grispy

# Data munging function
def position(row):
    return row[["tx", "ty", "tz"]].values.astype('float64')

# df is the big data set
df = pd.read_csv('test_data.csv')
gsp = grispy(df[["x", "y", "z"]].to_numpy().astype('float64'), copy_data=True)

# nvm data set has the points we want to do bubble neighbors to (camera positions)
# small bit of data munging here, as well as the vstack outside of grispy call
nvm_df = pd.read_csv('nvm_data.csv')
nvm_df['positions'] = nvm_df.apply(position, axis=1)
position_stack = np.vstack(nvm_df['positions'])


# Loop version, can use htop or similar utility to watch memory grow and then fall in each loop
# matching_indices = pd.Series()
step_size=300
envelope_size=8
for i in count(start=0, step=step_size):
    if i >= len(nvm_df['positions']):
        break
    distances, indices = gsp.bubble_neighbors(position_stack[i:i+step_size],
                                              distance_upper_bound=envelope_size)
    # flat_indices = list(chain.from_iterable(indices))
    # matching_indices = matching_indices.append(pd.Series(flat_indices))
# matching_indices = list(matching_indices.sort_values().unique())

```
The step_size variable will modify how many centers we're going through in a given loop, so if you set it appropriately (some 10s for a small memory-use iteration, some hundreds for an iteration that will eat 10s of GBs of memory for me), you can see the behavior I'm describing. The commented out code bits in the end of the loop / right out side of it, I figured might be useful to see the workaround that I end up using.  

Thanks again, and let me know if this script/data combo is useful / whether you can also see the issue. I appreciate the help!
Cheers,
Miguel

**Comment by mchalela** [comment]
Hi Miguel! Thank you for the detailed answer and for providing the test data. I downloaded both files and indeed my pc crashed. I'm not sure exactly why as I haven't made a memory profiling... something we will have to do soon. However, I was working on other pending issues and those fixes made your script work! I think it may have been the fix on issue #5 that did the trick.
The new version (version=0.1.0) is now on pypi, you can install it with `pip install grispy==0.1.0`.

This version of your script is now working fine on my pc:

```python
from itertools import count, chain
import numpy as np
import pandas as pd
from grispy import GriSPy as grispy

# Data munging function
def position(row):
    return row[["tx", "ty", "tz"]].values.astype('float64')

# df is the big data set
df = pd.read_csv('test_data.csv')
gsp = grispy(df[["x", "y", "z"]].to_numpy().astype('float64'), copy_data=True)

# nvm data set has the points we want to do bubble neighbors to (camera positions)
# small bit of data munging here, as well as the vstack outside of grispy call
nvm_df = pd.read_csv('nvm_data.csv')
nvm_df['positions'] = nvm_df.apply(position, axis=1)
position_stack = np.vstack(nvm_df['positions'])

# I removed the loop
distances, indices = gsp.bubble_neighbors(position_stack, distance_upper_bound=8.)
flat_indices = list(chain.from_iterable(indices))
matching_indices = pd.Series(flat_indices)
matching_indices = list(matching_indices.sort_values().unique())
```

Please, Miguel, let me know if the new version fixed your problem. If you still are having the same issue I'll make sure I run a memory profile on the code.

PD: A few side notes that may help with efficiency. When you cast your arrays with `.astype('float64')` this will by default return a copy of the array, so there should be no need for setting copy_data=True. Also, when you sort your indices with `matching_indices.sort_values().unique()`, you may see some speedup if you first get the unique values and then sort them.

**Comment by MLucient** [comment]
Hi Martin - thanks for the update. I upgraded GriSPy versions and can confirm that that seems to have fixed the issue, so I'll close here. Thanks also for the tips for my bit of code!
